core/src/main/scala/com/mesosphere/usi/core/scalaapi/SchedulerAdapter.scala

```diff
@@ -0,0 +1,170 @@
+    }
+
+    Source.maybe.prepend {
+      val events = Source
+        .unfoldResourceAsync[SpecUpdated, SinkQueueWithCancel[SpecUpdated]](
+          create = () => Future.successful(specQueue),
+          read = queue => queue.pull(),
+          close = queue =>
+            Future.successful {
+              killSwitch.shutdown()
+              queue.cancel()
+              Done
+          })
+        .watchTermination() {
+          case (_, completionSignal) =>
+            completionSignal.onComplete {
+              case Success(_) =>
+                killSwitch.shutdown()
+              case Failure(cause) =>
+                killSwitch.abort(cause)
+            }
+        }
+
+      Source.single(specsSnapshot -> events)
+    }.via(killSwitch.flow)
+      .via(schedulerFlow)
+      .flatMapConcat {
+        case (snapshot, updates) =>
+          stateSnapshotPromise.trySuccess(snapshot)
+          updates.watchTermination() {
+            case (_, cancellationSignal) =>
+              cancellationSignal.onComplete {
+                case Success(_) =>
+                  killSwitch.shutdown()
+                case Failure(cause) =>
+                  killSwitch.abort(cause)
+              }
+          }
+      }
+      .via(killSwitch.flow)
+      .mapAsync(1)(stateQueue.offer)
+      .map {
+        case QueueOfferResult.Enqueued =>
+        case QueueOfferResult.QueueClosed =>
+          killSwitch.shutdown()
+        case QueueOfferResult.Failure(cause) =>
+          killSwitch.abort(cause)
+        case QueueOfferResult.Dropped => // we shouldn't receive that at all because OverflowStrategy.backpressure
+          throw new RuntimeException("Unexpected QueueOfferResult.Dropped element")
+      }
+      .runWith(Sink.ignore)
+
+    val sourceWithKillSwitch = stateSource
+      .watchTermination() {
+        case (materializedValue, cancellationSignal) =>
+          cancellationSignal.onComplete {
+            case Success(_) =>
+              killSwitch.shutdown()
+            case Failure(cause) =>
+              killSwitch.abort(cause)
+          }
+          materializedValue
+      }
+      .via(killSwitch.flow)
+
+    val sinkWithKillSwitch = Flow[SpecUpdated]
+      .watchTermination() {
+        case (materializedValue, cancellationSignal) =>
+          cancellationSignal.onComplete {
+            case Success(_) =>
+              killSwitch.shutdown()
+            case Failure(cause) =>
+              killSwitch.abort(cause)
+          }
+          materializedValue
+      }
+      .via(killSwitch.flow)
+      .to(specSink)
+
+    (stateSnapshotPromise.future, sourceWithKillSwitch, sinkWithKillSwitch)
+
+  }
+
+  /**
+    * Represents the scheduler as a Flow.
+    *
+    * This method will materialize the scheduler first, then Flow can be materialized independently.
+    * @param specsSnapshot Snapshot of the current specs
+    * @return
+    */
+  def asFlow(specsSnapshot: SpecsSnapshot = SpecsSnapshot.empty)
```

Comment:
what would be the case where I would rather use this one? Is there one? If so, can we document it?

Otherwise I would rather have one recommended API than 10. Especially since one can create flow from the sink and source one so easily...

Reply:
Yeah, looks like a rare interface. I'll get rid of it!